BC break typescript >=2.8

After upgrading to vega@4.0.0:

`ERROR in node_modules/vega-typings/types/spec/util.d.ts(8,50): error TS2304: Cannot find name 'Exclude'.`

My app uses Angular 6 which is still at Typescript@2.7.2.
I guess **Exclude** was added in Typescript@2.8, so all apps below 2.8 will throw error.

--- a/types/spec/axis.d.ts
+++ b/types/spec/axis.d.ts
@@ -7,13 +7,28 @@ import {
   SignalRef,
   TextEncodeEntry,
 } from '.';
-import { FontWeight, Align, TextBaseline } from './encode';
+import { FontWeight, Align, TextBaseline, ArrayValueRef } from './encode';
+import {
+  NumberValue,
+  StringValue,
+  ColorValue,
+  FontWeightValue,
+  AlignValue,
+  TextBaselineValue,
+  SymbolShapeValue,
+  BooleanValue,
+} from './values';
+import { LayoutAlign } from './layout';
 
 export type AxisOrient = 'top' | 'bottom' | 'left' | 'right';
 
+export type LabelOverlap = boolean | 'parity' | 'greedy';
+
 export interface Axis extends BaseAxis {
   /**
-   * The orientation of the axis.
+   * The orientation of the axis. One of `"top"`, `"bottom"`, `"left"` or `"right"`. The orientation can be used to further specialize the axis type (e.g., a y axis oriented for the right edge of the chart).
+   *
+   * __Default value:__ `"bottom"` for x-axes and `"left"` for y-axes.
    */
   orient: AxisOrient;
 
@@ -28,14 +43,30 @@ export interface Axis extends BaseAxis {
   format?: string | SignalRef;
 
   /**
-   * The orthogonal offset in pixels by which to displace the axis from its position along the edge of the chart.
+   * A title for the axis (none by default).
    */
-  offset?: number | NumericValueRef;
+  title?: StringValue;
 
   /**
-   * The anchor position of the axis in pixels (default 0). For x-axes with top or bottom orientation, this sets the axis group x coordinate. For y-axes with left or right orientation, this sets the axis group y coordinate.
+   * The orthogonal offset in pixels by which to displace the axis from its position along the edge of the chart.
    */
-  position?: number | NumericValueRef;
+  offset?: NumberValue;
+
+  /**
+   * The anchor position of the axis in pixels. For x-axes with top or bottom orientation, this sets the axis group x coordinate. For y-axes with left or right orientation, this sets the axis group y coordinate.
+   *
+   * __Default value__: `0`
+   */
+  position?: NumberValue;
+
+  /**
+   * A desired number of ticks, for axes visualizing quantitative scales. The resulting number may be different so that values are "nice" (multiples of `2`, `5`, `10`) and lie within the underlying scale's range.
+   *
+   * For scales of type `"time"` or `"utc"`, the tick count can instead be a time interval specifier. Legal string values are `"millisecond"`, `"second"`, `"minute"`, `"hour"`, `"day"`, `"week"`, `"month"`, and "year". Alternatively, an object-valued interval specifier of the form `{"interval": "month", "step": 3}` includes a desired number of interval steps. Here, ticks are generated for each quarter (Jan, Apr, Jul, Oct) boundary.
+   *
+   * @minimum 0
+   */
+  tickCount?: number | TimeInterval;
 
   /**
    * Explicitly set the visible axis tick and label values.
@@ -83,93 +114,107 @@ export interface AxisEncode {
   domain?: GuideEncodeEntry<RuleEncodeEntry>;
 }
 
-export interface BaseAxis {
+export interface BaseAxis<
+  N = NumberValue,
+  NS = number | SignalRef,
+  B = BooleanValue,
+  BNS = number | boolean | SignalRef,
+  S = StringValue,
+  C = ColorValue,
+  FW = FontWeightValue,
+  A = AlignValue,
+  TB = TextBaselineValue,
+  LA = LayoutAlign | SignalRef,
+  SY = SymbolShapeValue,
+  LO = LabelOverlap | SignalRef,
+  DA = number[] | ArrayValueRef
+> {
   /**
    * The minimum extent in pixels that axis ticks and labels should use. This determines a minimum offset value for axis titles.
    *
    * __Default value:__ `30` for y-axis; `undefined` for x-axis.
    */
-  minExtent?: number | NumericValueRef;
+  minExtent?: N;
 
   /**
    * The maximum extent in pixels that axis ticks and labels should use. This determines a maximum offset value for axis titles.
    *
    * __Default value:__ `undefined`.
    */
-  maxExtent?: number | NumericValueRef;
+  maxExtent?: N;
 
   /**
    * An interpolation fraction indicating where, for `band` scales, axis ticks should be positioned. A value of `0` places ticks at the left edge of their bands. A value of `0.5` places ticks in the middle of their bands.
    *
    *  __Default value:__ `0.5`
    */
-  bandPosition?: number;
+  bandPosition?: N;
 
   // ---------- Title ----------
   /**
-   * A title for the axis (none by default).
-   */
-  title?: string | SignalRef;
-
-  /**
    * The padding, in pixels, between title and axis.
    */
-  titlePadding?: number | NumericValueRef;
+  titlePadding?: N;
 
   /**
    * Horizontal text alignment of axis titles.
    */
-  titleAlign?: Align;
+  titleAlign?: A;
 
   /**
    * Angle in degrees of axis titles.
    */
-  titleAngle?: number;
+  titleAngle?: N;
 
   /**
    * X-coordinate of the axis title relative to the axis group.
    */
-  titleX?: number;
+  titleX?: N;
 
   /**
    * Y-coordinate of the axis title relative to the axis group.
    */
-  titleY?: number;
+  titleY?: N;
 
   /**
    * Vertical text baseline for axis titles.
    */
-  titleBaseline?: TextBaseline;
+  titleBaseline?: TB;
 
   /**
    * Color of the title, can be in hex color code or regular color name.
    */
-  titleColor?: string;
+  titleColor?: C;
 
   /**
    * Font of the title. (e.g., `"Helvetica Neue"`).
    */
-  titleFont?: string;
+  titleFont?: S;
 
   /**
    * Font size of the title.
    *
    * @minimum 0
    */
-  titleFontSize?: number;
+  titleFontSize?: N;
 
   /**
    * Font weight of the title.
    * This can be either a string (e.g `"bold"`, `"normal"`) or a number (`100`, `200`, `300`, ..., `900` where `"normal"` = `400` and `"bold"` = `700`).
    */
-  titleFontWeight?: FontWeight;
+  titleFontWeight?: FW;
 
   /**
    * Maximum allowed pixel width of axis titles.
    *
    * @minimum 0
    */
-  titleLimit?: number;
+  titleLimit?: N;
+
+  /**
+   * Opacity of the axis title.
+   */
+  titleOpacity?: N;
 
   // ---------- Domain ----------
   /**
@@ -184,14 +229,19 @@ export interface BaseAxis {
    *
    * __Default value:__  `"gray"`.
    */
-  domainColor?: string;
+  domainColor?: C;
+
+  /**
+   * Opacity of the axis domain line.
+   */
+  domainOpacity?: N;
 
   /**
    * Stroke width of axis domain line
    *
    * __Default value:__  `1`
    */
-  domainWidth?: number;
+  domainWidth?: N;
 
   // ---------- Ticks ----------
   /**
@@ -199,33 +249,36 @@ export interface BaseAxis {
    *
    * __Default value:__ `true`
    */
-  ticks?: boolean;
-
-  tickCount?: number | TimeInterval;
+  ticks?: B;
 
   /**
    * The color of the axis's tick.
    *
    * __Default value:__ `"gray"`
    */
-  tickColor?: string;
+  tickColor?: C;
 
   /**
    * Boolean flag indicating if an extra axis tick should be added for the initial position of the axis. This flag is useful for styling axes for `band` scales such that ticks are placed on band boundaries rather in the middle of a band. Use in conjunction with `"bandPostion": 1` and an axis `"padding"` value of `0`.
    */
-  tickExtra?: boolean;
+  tickExtra?: B;
 
   /**
    * Position offset in pixels to apply to ticks, labels, and gridlines.
    */
-  tickOffset?: number;
+  tickOffset?: N;
+
+  /**
+   * Opacity of the ticks.
+   */
+  tickOpacity?: N;
 
   /**
    * Boolean flag indicating if pixel position values should be rounded to the nearest integer.
    *
    * __Default value:__ `true`
    */
-  tickRound?: boolean;
+  tickRound?: B;
 
   /**
    * The size in pixels of axis ticks.
@@ -233,7 +286,7 @@ export interface BaseAxis {
    * __Default value:__ `5`
    * @minimum 0
    */
-  tickSize?: number;
+  tickSize?: N;
 
   /**
    * The width, in pixels, of ticks.
@@ -241,13 +294,11 @@ export interface BaseAxis {
    * __Default value:__ `1`
    * @minimum 0
    */
-  tickWidth?: number;
+  tickWidth?: N;
 
   // ---------- Grid ----------
   /**
-   * A boolean flag indicating if grid lines should be included as part of the axis
-   *
-   * __Default value:__ `true` for [continuous scales](scale.html#continuous) that are not binned; otherwise, `false`.
+   * A boolean flag indicating if grid lines should be included as part of the axis.
    */
   grid?: boolean;
 
@@ -261,12 +312,12 @@ export interface BaseAxis {
    *
    * __Default value:__  `"lightGray"`.
    */
-  gridColor?: string;
+  gridColor?: C;
 
   /**
    * The offset (in pixels) into which to begin drawing with the grid dash array.
    */
-  gridDash?: number[];
+  gridDash?: DA;
 
   /**
    * The stroke opacity of grid (value between [0,1])
@@ -275,7 +326,7 @@ export interface BaseAxis {
    * @minimum 0
    * @maximum 1
    */
-  gridOpacity?: number;
+  gridOpacity?: N;
 
   /**
    * The grid width, in pixels.
@@ -283,7 +334,7 @@ export interface BaseAxis {
    * __Default value:__ `1`
    * @minimum 0
    */
-  gridWidth?: number;
+  gridWidth?: N;
 
   // ---------- Labels ----------
   /**
@@ -296,40 +347,36 @@ export interface BaseAxis {
   /**
    * Horizontal text alignment of axis tick labels, overriding the default setting for the current axis orientation.
    */
-  labelAlign?: Align;
+  labelAlign?: A;
 
   /**
    * Vertical text baseline of axis tick labels, overriding the default setting for the current axis orientation.
    */
-  labelBaseline?: TextBaseline;
+  labelBaseline?: TB;
 
   /**
    * Indicates if labels should be hidden if they exceed the axis range. If `false `(the default) no bounds overlap analysis is performed. If `true`, labels will be hidden if they exceed the axis range by more than 1 pixel. If this property is a number, it specifies the pixel tolerance: the maximum amount by which a label bounding box may exceed the axis range.
    *
    * __Default value:__ `false`.
    */
-  labelBound?: boolean | number;
+  labelBound?: BNS;
 
   /**
    * Indicates if the first and last axis labels should be aligned flush with the scale range. Flush alignment for a horizontal axis will left-align the first label and right-align the last label. For vertical axes, bottom and top text baselines are applied instead. If this property is a number, it also indicates the number of pixels by which to offset the first and last labels; for example, a value of 2 will flush-align the first and last labels and also push them 2 pixels outward from the center of the axis. The additional adjustment can sometimes help the labels better visually group with corresponding axis ticks.
-   *
-   * __Default value:__ `true` for axis of a continuous x-scale. Otherwise, `false`.
    */
-  labelFlush?: boolean | number;
+  labelFlush?: BNS;
 
   /**
    * Indicates the number of pixels by which to offset flush-adjusted labels. For example, a value of `2` will push flush-adjusted labels 2 pixels outward from the center of the axis. Offsets can help the labels better visually group with corresponding axis ticks.
    *
    * __Default value:__ `0`.
    */
-  labelFlushOffset?: number;
+  labelFlushOffset?: NS;
 
   /**
    * The strategy to use for resolving overlap of axis labels. If `false` (the default), no overlap reduction is attempted. If set to `true` or `"parity"`, a strategy of removing every other label is used (this works well for standard linear axes). If set to `"greedy"`, a linear scan of the labels is performed, removing any labels that overlaps with the last visible label (this often works better for log-scaled axes).
-   *
-   * __Default value:__ `true` for non-nominal fields with non-log scales; `"greedy"` for log scales; otherwise `false`.
    */
-  labelOverlap?: boolean | 'parity' | 'greedy';
+  labelOverlap?: LO;
 
   /**
    * The rotation angle of the axis labels.
@@ -339,36 +386,41 @@ export interface BaseAxis {
    * @minimum -360
    * @maximum 360
    */
-  labelAngle?: number;
+  labelAngle?: N;
 
   /**
    * The color of the tick label, can be in hex color code or regular color name.
    */
-  labelColor?: string;
+  labelColor?: C;
 
   /**
    * The font of the tick label.
    */
-  labelFont?: string;
+  labelFont?: S;
 
   /**
    * The font size of the label, in pixels.
    *
    * @minimum 0
    */
-  labelFontSize?: number;
+  labelFontSize?: N;
 
   /**
    * Font weight of axis tick labels.
    */
-  labelFontWeight?: FontWeight;
+  labelFontWeight?: FW;
 
   /**
    * Maximum allowed pixel width of axis tick labels.
    *
    * __Default value:__ `180`
    */
-  labelLimit?: number;
+  labelLimit?: N;
+
+  /**
+   * The opacity of the labels.
+   */
+  labelOpacity?: N;
 
   /**
    * The padding, in pixels, between axis and text labels.
@@ -376,5 +428,5 @@ export interface BaseAxis {
    * __Default value:__ `2`
    */
 
-  labelPadding?: number;
+  labelPadding?: N;
 }

--- a/types/spec/axis.d.ts
+++ b/types/spec/axis.d.ts
@@ -7,94 +7,378 @@ import {
   SignalRef,
   TextEncodeEntry,
 } from '.';
+import { FontWeight, Align, TextBaseline } from './encode';
+import { WithSignal } from './signal';
 
 export type AxisOrient = 'top' | 'bottom' | 'left' | 'right';
 
-export interface Axis {
+export interface Axis extends WithSignal<BaseAxis> {
+  /**
+   * The orientation of the axis.
+   */
   orient: AxisOrient;
+
+  /**
+   * The name of the scale backing the axis component.
+   */
   scale: string;
-  name?: string;
-  title?: string | SignalRef;
-  zindex?: number;
-  interactive?: boolean;
+
   /**
-   * Boolean value that determines whether the axis should include ticks.
+   * The format specifier pattern for axis labels. For numerical values, must be a legal [d3-format](https://github.com/d3/d3-format#locale_format) specifier. For date-time values, must be a legal [d3-time-format](https://github.com/d3/d3-time-format#locale_format) specifier.
    */
-  ticks?: boolean;
+  format?: string | SignalRef;
+
   /**
-   * A boolean flag indicating if labels should be included as part of the axis.
+   * The orthogonal offset in pixels by which to displace the axis from its position along the edge of the chart.
+   */
+
+  offset?: number | NumericValueRef;
+
+  /**
+   * The anchor position of the axis in pixels (default 0). For x-axes with top or bottom orientation, this sets the axis group x coordinate. For y-axes with left or right orientation, this sets the axis group y coordinate.
+   */
+
+  position?: number | NumericValueRef;
+
+  /**
+   * Explicitly set the visible axis tick and label values.
+   */
+
+  values?: any[] | SignalRef;
+
+  /**
+   * The integer z-index indicating the layering of the legend group relative to other axis, mark, and legend groups.
    *
-   * __Default value:__  `true`.
-   */
-  labels?: boolean;
-  /**
-   * Indicates if labels should be hidden if they exceed the axis range. If `false `(the default) no bounds overlap analysis is performed. If `true`, labels will be hidden if they exceed the axis range by more than 1 pixel. If this property is a number, it specifies the pixel tolerance: the maximum amount by which a label bounding box may exceed the axis range.
-   *
-   * __Default value:__ `false`.
-   */
-  labelBound?: boolean | number;
-  /**
-   * Indicates if the first and last axis labels should be aligned flush with the scale range. Flush alignment for a horizontal axis will left-align the first label and right-align the last label. For vertical axes, bottom and top text baselines are applied instead. If this property is a number, it also indicates the number of pixels by which to offset the first and last labels; for example, a value of 2 will flush-align the first and last labels and also push them 2 pixels outward from the center of the axis. The additional adjustment can sometimes help the labels better visually group with corresponding axis ticks.
-   *
-   * __Default value:__ `true` for axis of a continuous x-scale. Otherwise, `false`.
-   */
-  labelFlush?: boolean | number;
-  labelFlushOffset?: number;
-  /**
-   * The padding, in pixels, between axis and text labels.
-   */
-  labelPadding?: number;
-  /**
-   * The strategy to use for resolving overlap of axis labels. If `false` (the default), no overlap reduction is attempted. If set to `true` or `"parity"`, a strategy of removing every other label is used (this works well for standard linear axes). If set to `"greedy"`, a linear scan of the labels is performed, removing any labels that overlaps with the last visible label (this often works better for log-scaled axes).
-   *
-   * __Default value:__ `true` for non-nominal fields with non-log scales; `"greedy"` for log scales; otherwise `false`.
-   */
-  labelOverlap?: boolean | 'parity' | 'greedy';
-  /**
-   * A boolean flag indicating if the domain (the axis baseline) should be included as part of the axis.
-   *
-   * __Default value:__ `true`
-   */
-  domain?: boolean;
-  /**
-   * A boolean flag indicating if grid lines should be included as part of the axis
-   *
-   * __Default value:__ `true` for [continuous scales](scale.html#continuous) that are not binned; otherwise, `false`.
-   */
-  grid?: boolean;
-  gridScale?: string;
-  /**
-   * The size in pixels of axis ticks.
-   *
+   * @TJS-type integer
    * @minimum 0
    */
-  tickSize?: number;
-  tickCount?: number | TimeInterval | SignalRef;
-  format?: string | SignalRef;
-  values?: any[] | SignalRef;
-  offset?: number | NumericValueRef;
-  position?: number | NumericValueRef;
+  zindex?: number;
+
   /**
-   * The padding, in pixels, between title and axis.
+   * Mark definitions for custom axis encoding.
    */
-  titlePadding?: number | NumericValueRef;
+  encode?: AxisEncode;
+}
+
+export interface AxisEncode {
+  /**
+   * Custom encoding for the axis container.
+   */
+  axis?: GuideEncodeEntry<GroupEncodeEntry>;
+  /**
+   * Custom encoding for axis tick rule marks.
+   */
+  ticks?: GuideEncodeEntry<GroupEncodeEntry>;
+  /**
+   * Custom encoding for axis label text marks.
+   */
+  labels?: GuideEncodeEntry<TextEncodeEntry>;
+  /**
+   * Custom encoding for the axis title text mark.
+   */
+  title?: GuideEncodeEntry<TextEncodeEntry>;
+  /**
+   * Custom encoding for axis gridline rule marks.
+   */
+  grid?: GuideEncodeEntry<RuleEncodeEntry>;
+  /**
+   * Custom encoding for the axis domain rule mark.
+   */
+  domain?: GuideEncodeEntry<RuleEncodeEntry>;
+}
+
+export interface BaseAxis {
   /**
    * The minimum extent in pixels that axis ticks and labels should use. This determines a minimum offset value for axis titles.
    *
    * __Default value:__ `30` for y-axis; `undefined` for x-axis.
    */
   minExtent?: number | NumericValueRef;
+
   /**
    * The maximum extent in pixels that axis ticks and labels should use. This determines a maximum offset value for axis titles.
    *
    * __Default value:__ `undefined`.
    */
   maxExtent?: number | NumericValueRef;
-  encode?: {
-    ticks?: GuideEncodeEntry<GroupEncodeEntry>;
-    labels?: GuideEncodeEntry<TextEncodeEntry>;
-    title?: GuideEncodeEntry<TextEncodeEntry>;
-    grid?: GuideEncodeEntry<RuleEncodeEntry>;
-    domain?: GuideEncodeEntry<RuleEncodeEntry>;
-  };
+
+  /**
+   * An interpolation fraction indicating where, for `band` scales, axis ticks should be positioned. A value of `0` places ticks at the left edge of their bands. A value of `0.5` places ticks in the middle of their bands.
+   *
+   *  __Default value:__ `0.5`
+   */
+  bandPosition?: number;
+
+  // ---------- Title ----------
+  /**
+   * A title for the axis (none by default).
+   */
+  title?: string | SignalRef;
+
+  /**
+   * The padding, in pixels, between title and axis.
+   */
+  titlePadding?: number | NumericValueRef;
+
+  /**
+   * Horizontal text alignment of axis titles.
+   */
+  titleAlign?: Align;
+
+  /**
+   * Angle in degrees of axis titles.
+   */
+  titleAngle?: number;
+
+  /**
+   * X-coordinate of the axis title relative to the axis group.
+   */
+  titleX?: number;
+
+  /**
+   * Y-coordinate of the axis title relative to the axis group.
+   */
+  titleY?: number;
+
+  /**
+   * Vertical text baseline for axis titles.
+   */
+  titleBaseline?: TextBaseline;
+
+  /**
+   * Color of the title, can be in hex color code or regular color name.
+   */
+  titleColor?: string;
+
+  /**
+   * Font of the title. (e.g., `"Helvetica Neue"`).
+   */
+  titleFont?: string;
+
+  /**
+   * Font size of the title.
+   *
+   * @minimum 0
+   */
+  titleFontSize?: number;
+
+  /**
+   * Font weight of the title.
+   * This can be either a string (e.g `"bold"`, `"normal"`) or a number (`100`, `200`, `300`, ..., `900` where `"normal"` = `400` and `"bold"` = `700`).
+   */
+  titleFontWeight?: FontWeight;
+
+  /**
+   * Maximum allowed pixel width of axis titles.
+   *
+   * @minimum 0
+   */
+  titleLimit?: number;
+
+  // ---------- Domain ----------
+  /**
+   * A boolean flag indicating if the domain (the axis baseline) should be included as part of the axis.
+   *
+   * __Default value:__ `true`
+   */
+  domain?: boolean;
+
+  /**
+   * Color of axis domain line.
+   *
+   * __Default value:__  `"gray"`.
+   */
+  domainColor?: string;
+
+  /**
+   * Stroke width of axis domain line
+   *
+   * __Default value:__  `1`
+   */
+  domainWidth?: number;
+
+  // ---------- Ticks ----------
+  /**
+   * Boolean value that determines whether the axis should include ticks.
+   *
+   * __Default value:__ `true`
+   */
+  ticks?: boolean;
+
+  tickCount?: number | TimeInterval;
+
+  /**
+   * The color of the axis's tick.
+   *
+   * __Default value:__ `"gray"`
+   */
+  tickColor?: string;
+
+  /**
+   * Boolean flag indicating if an extra axis tick should be added for the initial position of the axis. This flag is useful for styling axes for `band` scales such that ticks are placed on band boundaries rather in the middle of a band. Use in conjunction with `"bandPostion": 1` and an axis `"padding"` value of `0`.
+   */
+  tickExtra?: boolean;
+
+  /**
+   * Position offset in pixels to apply to ticks, labels, and gridlines.
+   */
+  tickOffset?: number;
+
+  /**
+   * Boolean flag indicating if pixel position values should be rounded to the nearest integer.
+   *
+   * __Default value:__ `true`
+   */
+  tickRound?: boolean;
+
+  /**
+   * The size in pixels of axis ticks.
+   *
+   * __Default value:__ `5`
+   * @minimum 0
+   */
+  tickSize?: number;
+
+  /**
+   * The width, in pixels, of ticks.
+   *
+   * __Default value:__ `1`
+   * @minimum 0
+   */
+  tickWidth?: number;
+
+  // ---------- Grid ----------
+  /**
+   * A boolean flag indicating if grid lines should be included as part of the axis
+   *
+   * __Default value:__ `true` for [continuous scales](scale.html#continuous) that are not binned; otherwise, `false`.
+   */
+  grid?: boolean;
+
+  /**
+   * The name of the scale to use for including grid lines. By default grid lines are driven by the same scale as the ticks and labels.
+   */
+  gridScale?: string;
+
+  /**
+   * Color of gridlines.
+   *
+   * __Default value:__  `"lightGray"`.
+   */
+  gridColor?: string;
+
+  /**
+   * The offset (in pixels) into which to begin drawing with the grid dash array.
+   */
+  gridDash?: number[];
+
+  /**
+   * The stroke opacity of grid (value between [0,1])
+   *
+   * __Default value:__ `1`
+   * @minimum 0
+   * @maximum 1
+   */
+  gridOpacity?: number;
+
+  /**
+   * The grid width, in pixels.
+   *
+   * __Default value:__ `1`
+   * @minimum 0
+   */
+  gridWidth?: number;
+
+  // ---------- Labels ----------
+  /**
+   * A boolean flag indicating if labels should be included as part of the axis.
+   *
+   * __Default value:__  `true`.
+   */
+  labels?: boolean;
+
+  /**
+   * Horizontal text alignment of axis tick labels, overriding the default setting for the current axis orientation.
+   */
+  labelAlign?: Align;
+
+  /**
+   * Vertical text baseline of axis tick labels, overriding the default setting for the current axis orientation.
+   */
+  labelBaseline?: TextBaseline;
+
+  /**
+   * Indicates if labels should be hidden if they exceed the axis range. If `false `(the default) no bounds overlap analysis is performed. If `true`, labels will be hidden if they exceed the axis range by more than 1 pixel. If this property is a number, it specifies the pixel tolerance: the maximum amount by which a label bounding box may exceed the axis range.
+   *
+   * __Default value:__ `false`.
+   */
+  labelBound?: boolean | number;
+
+  /**
+   * Indicates if the first and last axis labels should be aligned flush with the scale range. Flush alignment for a horizontal axis will left-align the first label and right-align the last label. For vertical axes, bottom and top text baselines are applied instead. If this property is a number, it also indicates the number of pixels by which to offset the first and last labels; for example, a value of 2 will flush-align the first and last labels and also push them 2 pixels outward from the center of the axis. The additional adjustment can sometimes help the labels better visually group with corresponding axis ticks.
+   *
+   * __Default value:__ `true` for axis of a continuous x-scale. Otherwise, `false`.
+   */
+  labelFlush?: boolean | number;
+
+  /**
+   * Indicates the number of pixels by which to offset flush-adjusted labels. For example, a value of `2` will push flush-adjusted labels 2 pixels outward from the center of the axis. Offsets can help the labels better visually group with corresponding axis ticks.
+   *
+   * __Default value:__ `0`.
+   */
+  labelFlushOffset?: number;
+
+  /**
+   * The strategy to use for resolving overlap of axis labels. If `false` (the default), no overlap reduction is attempted. If set to `true` or `"parity"`, a strategy of removing every other label is used (this works well for standard linear axes). If set to `"greedy"`, a linear scan of the labels is performed, removing any labels that overlaps with the last visible label (this often works better for log-scaled axes).
+   *
+   * __Default value:__ `true` for non-nominal fields with non-log scales; `"greedy"` for log scales; otherwise `false`.
+   */
+  labelOverlap?: boolean | 'parity' | 'greedy';
+
+  /**
+   * The rotation angle of the axis labels.
+   *
+   * __Default value:__ `-90` for nominal and ordinal fields; `0` otherwise.
+   *
+   * @minimum -360
+   * @maximum 360
+   */
+  labelAngle?: number;
+
+  /**
+   * The color of the tick label, can be in hex color code or regular color name.
+   */
+  labelColor?: string;
+
+  /**
+   * The font of the tick label.
+   */
+  labelFont?: string;
+
+  /**
+   * The font size of the label, in pixels.
+   *
+   * @minimum 0
+   */
+  labelFontSize?: number;
+
+  /**
+   * Font weight of axis tick labels.
+   */
+  labelFontWeight?: FontWeight;
+
+  /**
+   * Maximum allowed pixel width of axis tick labels.
+   *
+   * __Default value:__ `180`
+   */
+  labelLimit?: number;
+
+  /**
+   * The padding, in pixels, between axis and text labels.
+   *
+   * __Default value:__ `2`
+   */
+
+  labelPadding?: number;
 }

--- a/types/spec/axis.d.ts
+++ b/types/spec/axis.d.ts
@@ -8,11 +8,10 @@ import {
   TextEncodeEntry,
 } from '.';
 import { FontWeight, Align, TextBaseline } from './encode';
-import { WithSignal } from './signal';
 
 export type AxisOrient = 'top' | 'bottom' | 'left' | 'right';
 
-export interface Axis extends WithSignal<BaseAxis> {
+export interface Axis extends BaseAxis {
   /**
    * The orientation of the axis.
    */
@@ -31,19 +30,16 @@ export interface Axis extends WithSignal<BaseAxis> {
   /**
    * The orthogonal offset in pixels by which to displace the axis from its position along the edge of the chart.
    */
-
   offset?: number | NumericValueRef;
 
   /**
    * The anchor position of the axis in pixels (default 0). For x-axes with top or bottom orientation, this sets the axis group x coordinate. For y-axes with left or right orientation, this sets the axis group y coordinate.
    */
-
   position?: number | NumericValueRef;
 
   /**
    * Explicitly set the visible axis tick and label values.
    */
-
   values?: any[] | SignalRef;
 
   /**

--- a/types/spec/config.d.ts
+++ b/types/spec/config.d.ts
@@ -13,7 +13,7 @@ import {
   TitleAnchor,
   TitleOrient,
 } from '.';
-import { BaseAxis } from './axis';
+import { BaseAxis, LabelOverlap } from './axis';
 import { BaseLegend } from './legend';
 import {
   NumberValue,
@@ -23,6 +23,7 @@ import {
   AlignValue,
   TextBaselineValue,
   SymbolShapeValue,
+  BooleanValue,
 } from './values';
 import { SignalRef } from './signal';
 import { LayoutAlign } from './layout';
@@ -337,14 +338,16 @@ export type AxisConfig = BaseAxis;
 export interface LegendConfig<
   N = NumberValue,
   NS = number | SignalRef,
+  B = BooleanValue,
   S = StringValue,
   C = ColorValue,
   FW = FontWeightValue,
   A = AlignValue,
-  B = TextBaselineValue,
+  TB = TextBaselineValue,
   LA = LayoutAlign | SignalRef,
+  LO = LabelOverlap | SignalRef,
   SY = SymbolShapeValue
-> extends BaseLegend<N, NS, S, C, FW, A, B, LA, SY> {
+> extends BaseLegend<N, NS, B, S, C, FW, A, TB, LA, LO, SY> {
   /**
    * The default direction (`"horizontal"` or `"vertical"`) for gradient legends.
    *

--- a/types/spec/config.d.ts
+++ b/types/spec/config.d.ts
@@ -1,7 +1,6 @@
 import {
   Align,
   AutoSize,
-  Axis,
   EventType,
   FontStyle,
   FontWeight,
@@ -14,8 +13,9 @@ import {
   TitleAnchor,
   TitleOrient,
 } from '.';
-import { BaseLegend, LegendDirection } from './legend';
-import { SignalRef, WithSignal } from './signal';
+import { BaseAxis } from './axis';
+import { BaseLegend } from './legend';
+import { WithSignal } from './signal';
 import { Omit } from './util';
 
 export interface Config
@@ -320,158 +320,9 @@ export type AxisConfigKeys =
   | 'axisLeft'
   | 'axisBand';
 
-export interface AxisConfig extends Partial<Axis> {
-  /**
-   * An interpolation fraction indicating where, for `band` scales, axis ticks should be positioned. A value of `0` places ticks at the left edge of their bands. A value of `0.5` places ticks in the middle of their bands.
-   */
-  bandPosition?: number;
-  /**
-   * Stroke width of axis domain line
-   *
-   * __Default value:__  (none, using Vega default).
-   */
-  domainWidth?: number;
+export interface BaseAxisConfig extends BaseAxis {}
 
-  /**
-   * Color of axis domain line.
-   *
-   * __Default value:__  (none, using Vega default).
-   */
-  domainColor?: string;
-
-  // ---------- Grid ----------
-  /**
-   * Color of gridlines.
-   */
-  gridColor?: string;
-
-  /**
-   * The offset (in pixels) into which to begin drawing with the grid dash array.
-   */
-  gridDash?: number[];
-
-  /**
-   * The stroke opacity of grid (value between [0,1])
-   *
-   * __Default value:__ (`1` by default)
-   * @minimum 0
-   * @maximum 1
-   */
-  gridOpacity?: number;
-
-  /**
-   * The grid width, in pixels.
-   * @minimum 0
-   */
-  gridWidth?: number;
-
-  // ---------- Ticks ----------
-  /**
-   * The color of the axis's tick.
-   */
-  tickColor?: string;
-
-  /**
-   * Boolean flag indicating if an extra axis tick should be added for the initial position of the axis. This flag is useful for styling axes for `band` scales such that ticks are placed on band boundaries rather in the middle of a band. Use in conjunction with `"bandPostion": 1` and an axis `"padding"` value of `0`.
-   */
-  tickExtra?: boolean;
-
-  /**
-   * The rotation angle of the axis labels.
-   *
-   * __Default value:__ `-90` for nominal and ordinal fields; `0` otherwise.
-   *
-   * @minimum -360
-   * @maximum 360
-   */
-  labelAngle?: number;
-
-  /**
-   * The color of the tick label, can be in hex color code or regular color name.
-   */
-  labelColor?: string;
-
-  /**
-   * The font of the tick label.
-   */
-  labelFont?: string;
-
-  /**
-   * The font size of the label, in pixels.
-   *
-   * @minimum 0
-   */
-  labelFontSize?: number;
-
-  /**
-   * Maximum allowed pixel width of axis tick labels.
-   */
-  labelLimit?: number;
-
-  /**
-   * Boolean flag indicating if pixel position values should be rounded to the nearest integer.
-   */
-  tickRound?: boolean;
-
-  /**
-   * The width, in pixels, of ticks.
-   *
-   * @minimum 0
-   */
-  tickWidth?: number;
-
-  // ---------- Title ----------
-  /**
-   * Horizontal text alignment of axis titles.
-   */
-  titleAlign?: string;
-
-  /**
-   * Angle in degrees of axis titles.
-   */
-  titleAngle?: number;
-  /**
-   * Vertical text baseline for axis titles.
-   */
-  titleBaseline?: string;
-  /**
-   * Color of the title, can be in hex color code or regular color name.
-   */
-  titleColor?: string;
-
-  /**
-   * Font of the title. (e.g., `"Helvetica Neue"`).
-   */
-  titleFont?: string;
-
-  /**
-   * Font size of the title.
-   *
-   * @minimum 0
-   */
-  titleFontSize?: number;
-
-  /**
-   * Font weight of the title.
-   * This can be either a string (e.g `"bold"`, `"normal"`) or a number (`100`, `200`, `300`, ..., `900` where `"normal"` = `400` and `"bold"` = `700`).
-   */
-  titleFontWeight?: FontWeight;
-
-  /**
-   * Maximum allowed pixel width of axis titles.
-   */
-  titleLimit?: number;
-
-  /**
-   * X-coordinate of the axis title relative to the axis group.
-   */
-  titleX?: number;
-
-  /**
-   * Y-coordinate of the axis title relative to the axis group.
-   */
-  titleY?: number;
-}
+export type AxisConfig = WithSignal<BaseAxisConfig>;
 
 /**
  * Legend Config without signals so we can use it in Vega-Lite.
@@ -482,7 +333,7 @@ export interface BaseLegendConfig extends BaseLegend {
    *
    * __Default value:__ `"vertical"`.
    */
-  gradientDirection?: LegendDirection;
+  gradientDirection?: Orientation;
 
   /**
    * The maximum allowed length in pixels of color ramp gradient labels.
@@ -515,7 +366,7 @@ export interface BaseLegendConfig extends BaseLegend {
    *
    * __Default value:__ `"vertical"`.
    */
-  symbolDirection?: LegendDirection;
+  symbolDirection?: Orientation;
 
   /**
    * Horizontal pixel offset for legend symbols.

--- a/types/spec/config.d.ts
+++ b/types/spec/config.d.ts
@@ -15,8 +15,17 @@ import {
 } from '.';
 import { BaseAxis } from './axis';
 import { BaseLegend } from './legend';
-import { WithSignal } from './signal';
-import { Omit } from './util';
+import {
+  NumberValue,
+  StringValue,
+  ColorValue,
+  FontWeightValue,
+  AlignValue,
+  TextBaselineValue,
+  SymbolShapeValue,
+} from './values';
+import { SignalRef } from './signal';
+import { LayoutAlign } from './layout';
 
 export interface Config
   extends Partial<Record<MarkConfigKeys, MarkConfig>>,
@@ -320,14 +329,22 @@ export type AxisConfigKeys =
   | 'axisLeft'
   | 'axisBand';
 
-export interface BaseAxisConfig extends BaseAxis {}
-
-export type AxisConfig = WithSignal<BaseAxisConfig>;
+export type AxisConfig = BaseAxis;
 
 /**
  * Legend Config without signals so we can use it in Vega-Lite.
  */
-export interface BaseLegendConfig extends BaseLegend {
+export interface LegendConfig<
+  N = NumberValue,
+  NS = number | SignalRef,
+  S = StringValue,
+  C = ColorValue,
+  FW = FontWeightValue,
+  A = AlignValue,
+  B = TextBaselineValue,
+  LA = LayoutAlign | SignalRef,
+  SY = SymbolShapeValue
+> extends BaseLegend<N, NS, S, C, FW, A, B, LA, SY> {
   /**
    * The default direction (`"horizontal"` or `"vertical"`) for gradient legends.
    *
@@ -338,28 +355,28 @@ export interface BaseLegendConfig extends BaseLegend {
   /**
    * The maximum allowed length in pixels of color ramp gradient labels.
    */
-  gradientLabelLimit?: number;
+  gradientLabelLimit?: N;
 
   /**
    * Vertical offset in pixels for color ramp gradient labels.
    *
    * __Default value:__ `2`.
    */
-  gradientLabelOffset?: number;
+  gradientLabelOffset?: N;
 
   /**
    * Default fill color for legend symbols. Only applied if there is no `"fill"` scale color encoding for the legend.
    *
    * __Default value:__ `"transparent"`.
    */
-  symbolBaseFillColor?: string;
+  symbolBaseFillColor?: C;
 
   /**
    * Default stroke color for legend symbols. Only applied if there is no `"fill"` scale color encoding for the legend.
    *
    * __Default value:__ `"gray"`.
    */
-  symbolBaseStrokeColor?: string;
+  symbolBaseStrokeColor?: C;
 
   /**
    * The default direction (`"horizontal"` or `"vertical"`) for symbol legends.
@@ -369,21 +386,10 @@ export interface BaseLegendConfig extends BaseLegend {
   symbolDirection?: Orientation;
 
   /**
-   * Horizontal pixel offset for legend symbols.
-   *
-   * __Default value:__ `0`.
-   */
-  symbolOffset?: number;
-
-  /**
    * Border stroke dash pattern for the full legend.
    */
   strokeDash?: number[];
 }
-
-export interface LegendConfig
-  extends WithSignal<Omit<BaseLegendConfig, 'orient'>>,
-    Pick<BaseLegendConfig, 'orient'> {}
 
 export interface TitleConfig {
   /**

--- a/types/spec/encode.d.ts
+++ b/types/spec/encode.d.ts
@@ -46,6 +46,7 @@ export type FontWeightValueRef = ScaledValueRef<FontWeight>;
 export type AlignValueRef = ScaledValueRef<Align>;
 export type TextBaselineValueRef = ScaledValueRef<TextBaseline>;
 export type BooleanValueRef = ScaledValueRef<boolean>;
+export type ArrayValueRef = ScaledValueRef<any[]>;
 export type ArbitraryValueRef = NumericValueRef | ColorValueRef | ScaledValueRef<any>;
 export interface ColorRGB {
   r: NumericValueRef;

--- a/types/spec/encode.d.ts
+++ b/types/spec/encode.d.ts
@@ -41,6 +41,10 @@ export type NumericValueRef = (ScaledValueRef<number> | {}) & {
   extra?: boolean;
 };
 export type StringValueRef = ScaledValueRef<string>;
+export type SymbolShapeValueRef = ScaledValueRef<SymbolShape>;
+export type FontWeightValueRef = ScaledValueRef<FontWeight>;
+export type AlignValueRef = ScaledValueRef<Align>;
+export type TextBaselineValueRef = ScaledValueRef<TextBaseline>;
 export type BooleanValueRef = ScaledValueRef<boolean>;
 export type ArbitraryValueRef = NumericValueRef | ColorValueRef | ScaledValueRef<any>;
 export interface ColorRGB {

--- a/types/spec/legend.d.ts
+++ b/types/spec/legend.d.ts
@@ -6,9 +6,10 @@ import {
   SymbolEncodeEntry,
   TextEncodeEntry,
 } from '.';
-import { FontWeight, SymbolShape } from './encode';
+import { FontWeight, SymbolShape, TextBaseline } from './encode';
 import { WithSignal } from './signal';
 import { Omit } from './util';
+import { TimeInterval } from './scale';
 
 export interface GuideEncodeEntry<T> {
   name?: string;
@@ -33,8 +34,6 @@ export type LegendOrient =
   | 'bottom-left'
   | 'bottom-right';
 
-export type LegendDirection = 'vertical' | 'horizontal';
-
 export interface Legend extends WithSignal<Omit<BaseLegend, 'orient'>>, Pick<BaseLegend, 'orient'> {
   size?: string;
   shape?: string;
@@ -53,7 +52,7 @@ export interface Legend extends WithSignal<Omit<BaseLegend, 'orient'>>, Pick<Bas
    *
    * __Default value:__ `"vertical"`
    */
-  direction?: LegendDirection;
+  direction?: Orientation;
 
   /**
    * The format specifier pattern for legend labels. For numerical values, must be a legal [d3-format](https://github.com/d3/d3-format#locale_format) specifier. For date-time values, must be a legal [d3-time-format](https://github.com/d3/d3-time-format#locale_format) specifier.
@@ -68,7 +67,7 @@ export interface Legend extends WithSignal<Omit<BaseLegend, 'orient'>>, Pick<Bas
   /**
    * The desired number of tick values for quantitative legends.
    */
-  tickCount?: number | SignalRef;
+  tickCount?: number | TimeInterval;
 
   /**
    * Explicitly set the visible legend values.
@@ -76,7 +75,7 @@ export interface Legend extends WithSignal<Omit<BaseLegend, 'orient'>>, Pick<Bas
   values?: any[] | SignalRef;
 
   /**
-   * The integer z-index indicating the layering of the legend group relative to other axis, mark and legend groups.
+   * The integer z-index indicating the layering of the legend group relative to other axis, mark, and legend groups.
    *
    * @TJS-type integer
    * @minimum 0
@@ -92,7 +91,8 @@ export interface Legend extends WithSignal<Omit<BaseLegend, 'orient'>>, Pick<Bas
 export interface LegendEncode {
   title?: GuideEncodeEntry<GroupEncodeEntry>;
   labels?: GuideEncodeEntry<TextEncodeEntry>;
-  legend?: GuideEncodeEntry<TextEncodeEntry>;
+  legend?: GuideEncodeEntry<GroupEncodeEntry>;
+  entries?: GuideEncodeEntry<TextEncodeEntry>;
   symbols?: GuideEncodeEntry<SymbolEncodeEntry>;
   gradient?: GuideEncodeEntry<RectEncodeEntry>;
 }
@@ -156,7 +156,7 @@ export interface BaseLegend {
    *
    * __Default value:__ `"top"`.
    */
-  titleBaseline?: string;
+  titleBaseline?: TextBaseline;
   /**
    * The color of the legend title, can be in hex color code or regular color name.
    */
@@ -294,16 +294,16 @@ export interface BaseLegend {
 
   // ---------- Label ----------
   /**
-   * The alignment of the legend label, can be left, middle or right.
+   * The alignment of the legend label, can be left, center, or right.
    */
   labelAlign?: string;
 
   /**
-   * The position of the baseline of legend label, can be top, middle or bottom.
+   * The position of the baseline of legend label, can be top, middle, bottom, or alphabetic.
    *
    * __Default value:__ `"middle"`.
    */
-  labelBaseline?: string;
+  labelBaseline?: TextBaseline;
 
   /**
    * The color of the legend label, can be in hex color code or regular color name.

--- a/types/spec/legend.d.ts
+++ b/types/spec/legend.d.ts
@@ -16,8 +16,10 @@ import {
   TextBaselineValue,
   SymbolShapeValue,
   ColorValue,
+  BooleanValue,
 } from './values';
 import { LayoutAlign } from './layout';
+import { LabelOverlap } from './axis';
 
 export interface GuideEncodeEntry<T> {
   name?: string;
@@ -111,12 +113,14 @@ export interface LegendEncode {
 export interface BaseLegend<
   N = NumberValue,
   NS = number | SignalRef,
+  B = BooleanValue,
   S = StringValue,
   C = ColorValue,
   FW = FontWeightValue,
   A = AlignValue,
-  B = TextBaselineValue,
+  TB = TextBaselineValue,
   LA = LayoutAlign | SignalRef,
+  LO = LabelOverlap | SignalRef,
   SY = SymbolShapeValue
 > {
   /**
@@ -174,7 +178,7 @@ export interface BaseLegend<
    *
    * __Default value:__ `"top"`.
    */
-  titleBaseline?: B;
+  titleBaseline?: TB;
   /**
    * The color of the legend title, can be in hex color code or regular color name.
    */
@@ -205,6 +209,11 @@ export interface BaseLegend<
   titleLimit?: N;
 
   /**
+   * Opacity of the legend title.
+   */
+  titleOpacity?: N;
+
+  /**
    * The padding, in pixels, between title and legend.
    *
    * __Default value:__ `5`.
@@ -220,6 +229,11 @@ export interface BaseLegend<
    * @minimum 0
    */
   gradientLength?: NS;
+
+  /**
+   * Opacity of the color gradient.
+   */
+  gradientOpacity?: N;
 
   /**
    * The thickness in pixels of the color gradient. This value corresponds to the width of a vertical gradient or the height of a horizontal gradient.
@@ -242,7 +256,7 @@ export interface BaseLegend<
    * __Default value:__ `0`.
    * @minimum 0
    */
-  gradientStrokeWidth?: NS;
+  gradientStrokeWidth?: N;
 
   // ---------- Symbol Layout ----------
   /**
@@ -290,6 +304,11 @@ export interface BaseLegend<
   symbolOffset?: N;
 
   /**
+   * Opacity of the legend symbols.
+   */
+  symbolOpacity?: N;
+
+  /**
    * The size of the legend symbol, in pixels.
    *
    * __Default value:__ `100`.
@@ -328,7 +347,7 @@ export interface BaseLegend<
    *
    * __Default value:__ `"middle"`.
    */
-  labelBaseline?: B;
+  labelBaseline?: TB;
 
   /**
    * The color of the legend label, can be in hex color code or regular color name.
@@ -362,6 +381,16 @@ export interface BaseLegend<
   labelLimit?: N;
 
   /**
+   * Opacity of labels.
+   */
+  labelOpacity?: N;
+
+  /**
+   * Padding in pixels between the legend and legend labels.
+   */
+  labelPadding?: N;
+
+  /**
    * The offset of the legend label.
    * @minimum 0
    *
@@ -374,5 +403,5 @@ export interface BaseLegend<
    *
    * __Default value:__ `true`.
    */
-  labelOverlap?: boolean | 'parity' | 'greedy';
+  labelOverlap?: LabelOverlap;
 }

--- a/types/spec/legend.d.ts
+++ b/types/spec/legend.d.ts
@@ -6,10 +6,18 @@ import {
   SymbolEncodeEntry,
   TextEncodeEntry,
 } from '.';
-import { FontWeight, SymbolShape, TextBaseline } from './encode';
-import { WithSignal } from './signal';
-import { Omit } from './util';
+import { FontWeight, SymbolShape, TextBaseline, ColorValueRef } from './encode';
 import { TimeInterval } from './scale';
+import {
+  NumberValue,
+  StringValue,
+  AlignValue,
+  FontWeightValue,
+  TextBaselineValue,
+  SymbolShapeValue,
+  ColorValue,
+} from './values';
+import { LayoutAlign } from './layout';
 
 export interface GuideEncodeEntry<T> {
   name?: string;
@@ -34,7 +42,7 @@ export type LegendOrient =
   | 'bottom-left'
   | 'bottom-right';
 
-export interface Legend extends WithSignal<Omit<BaseLegend, 'orient'>>, Pick<BaseLegend, 'orient'> {
+export interface Legend extends BaseLegend {
   size?: string;
   shape?: string;
   fill?: string;
@@ -98,9 +106,19 @@ export interface LegendEncode {
 }
 
 /**
- * Properties shared between legends and legend configs. Without signals so we can use it in Vega-Lite.
+ * Properties shared between legends and legend configs.
  */
-export interface BaseLegend {
+export interface BaseLegend<
+  N = NumberValue,
+  NS = number | SignalRef,
+  S = StringValue,
+  C = ColorValue,
+  FW = FontWeightValue,
+  A = AlignValue,
+  B = TextBaselineValue,
+  LA = LayoutAlign | SignalRef,
+  SY = SymbolShapeValue
+> {
   /**
    * The orientation of the legend, which determines how the legend is positioned within the scene. One of "left", "right", "top-left", "top-right", "bottom-left", "bottom-right", "none".
    *
@@ -112,36 +130,36 @@ export interface BaseLegend {
   /**
    * Corner radius for the full legend.
    */
-  cornerRadius?: number;
+  cornerRadius?: N;
 
   /**
    * Background fill color for the full legend.
    */
-  fillColor?: string;
+  fillColor?: C;
 
   /**
    * The offset in pixels by which to displace the legend from the data rectangle and axes.
    *
    * __Default value:__ `18`.
    */
-  offset?: number;
+  offset?: N;
 
   /**
    * The padding between the border and content of the legend group.
    *
    * __Default value:__ `0`.
    */
-  padding?: number;
+  padding?: N;
 
   /**
    * Border stroke color for the full legend.
    */
-  strokeColor?: string;
+  strokeColor?: C;
 
   /**
    * Border stroke width for the full legend.
    */
-  strokeWidth?: number;
+  strokeWidth?: N;
 
   // ---------- Title ----------
   /**
@@ -149,34 +167,34 @@ export interface BaseLegend {
    *
    * __Default value:__ `"left"`.
    */
-  titleAlign?: string;
+  titleAlign?: A;
 
   /**
    * Vertical text baseline for legend titles.
    *
    * __Default value:__ `"top"`.
    */
-  titleBaseline?: TextBaseline;
+  titleBaseline?: B;
   /**
    * The color of the legend title, can be in hex color code or regular color name.
    */
-  titleColor?: string;
+  titleColor?: C;
 
   /**
    * The font of the legend title.
    */
-  titleFont?: string;
+  titleFont?: S;
 
   /**
    * The font size of the legend title.
    */
-  titleFontSize?: number;
+  titleFontSize?: N;
 
   /**
    * The font weight of the legend title.
    * This can be either a string (e.g `"bold"`, `"normal"`) or a number (`100`, `200`, `300`, ..., `900` where `"normal"` = `400` and `"bold"` = `700`).
    */
-  titleFontWeight?: FontWeight;
+  titleFontWeight?: FW;
 
   /**
    * Maximum allowed pixel width of axis titles.
@@ -184,14 +202,14 @@ export interface BaseLegend {
    * __Default value:__ `180`.
    * @minimum 0
    */
-  titleLimit?: number;
+  titleLimit?: N;
 
   /**
    * The padding, in pixels, between title and legend.
    *
    * __Default value:__ `5`.
    */
-  titlePadding?: number;
+  titlePadding?: N;
 
   // ---------- Gradient ----------
 
@@ -201,7 +219,7 @@ export interface BaseLegend {
    * __Default value:__ `200`.
    * @minimum 0
    */
-  gradientLength?: number;
+  gradientLength?: NS;
 
   /**
    * The thickness in pixels of the color gradient. This value corresponds to the width of a vertical gradient or the height of a horizontal gradient.
@@ -209,14 +227,14 @@ export interface BaseLegend {
    * __Default value:__ `16`.
    * @minimum 0
    */
-  gradientThickness?: number;
+  gradientThickness?: NS;
 
   /**
    * The color of the gradient stroke, can be in hex color code or regular color name.
    *
    * __Default value:__ `"lightGray"`.
    */
-  gradientStrokeColor?: string;
+  gradientStrokeColor?: C;
 
   /**
    * The width of the gradient stroke, in pixels.
@@ -224,45 +242,52 @@ export interface BaseLegend {
    * __Default value:__ `0`.
    * @minimum 0
    */
-  gradientStrokeWidth?: number;
+  gradientStrokeWidth?: NS;
 
   // ---------- Symbol Layout ----------
   /**
    * The height in pixels to clip symbol legend entries and limit their size.
    */
-  clipHeight?: number;
+  clipHeight?: NS;
 
   /**
    * The number of columns in which to arrange symbol legend entries. A value of `0` or lower indicates a single row with one column per entry.
    */
-  columns?: number;
+  columns?: NS;
 
   /**
    * The horizontal padding in pixels between symbol legend entries.
    *
    * __Default value:__ `10`.
    */
-  columnPadding?: number;
+  columnPadding?: NS;
 
   /**
    * The vertical padding in pixels between symbol legend entries.
    *
    * __Default value:__ `2`.
    */
-  rowPadding?: number;
+  rowPadding?: NS;
 
   /**
    * The alignment to apply to symbol legends rows and columns. The supported string values are `"all"`, `"each"` (the default), and `none`. For more information, see the [grid layout documentation](https://vega.github.io/vega/docs/layout).
    *
    * __Default value:__ `"each"`.
    */
-  gridAlign?: 'all' | 'each' | 'none';
+  gridAlign?: LA;
 
   // ---------- Symbols ----------
   /**
    * The color of the legend symbol,
    */
-  symbolFillColor?: string;
+  symbolFillColor?: C;
+
+  /**
+   * Horizontal pixel offset for legend symbols.
+   *
+   * __Default value:__ `0`.
+   */
+  symbolOffset?: N;
 
   /**
    * The size of the legend symbol, in pixels.
@@ -270,12 +295,12 @@ export interface BaseLegend {
    * __Default value:__ `100`.
    * @minimum 0
    */
-  symbolSize?: number;
+  symbolSize?: N;
 
   /**
    * Stroke color for legend symbols.
    */
-  symbolStrokeColor?: string;
+  symbolStrokeColor?: C;
 
   /**
    * The width of the symbol's stroke.
@@ -283,37 +308,37 @@ export interface BaseLegend {
    * __Default value:__ `1.5`.
    * @minimum 0
    */
-  symbolStrokeWidth?: number;
+  symbolStrokeWidth?: N;
 
   /**
    * Default shape type (such as "circle") for legend symbols.
    *
    * __Default value:__ `"circle"`.
    */
-  symbolType?: SymbolShape;
+  symbolType?: SY;
 
   // ---------- Label ----------
   /**
    * The alignment of the legend label, can be left, center, or right.
    */
-  labelAlign?: string;
+  labelAlign?: A;
 
   /**
    * The position of the baseline of legend label, can be top, middle, bottom, or alphabetic.
    *
    * __Default value:__ `"middle"`.
    */
-  labelBaseline?: TextBaseline;
+  labelBaseline?: B;
 
   /**
    * The color of the legend label, can be in hex color code or regular color name.
    */
-  labelColor?: string;
+  labelColor?: C;
 
   /**
    * The font of the legend label.
    */
-  labelFont?: string;
+  labelFont?: S;
 
   /**
    * The font size of legend label.
@@ -322,19 +347,19 @@ export interface BaseLegend {
    *
    * @minimum 0
    */
-  labelFontSize?: number;
+  labelFontSize?: N;
 
   /**
    * The font weight of legend label.
    */
-  labelFontWeight?: FontWeight;
+  labelFontWeight?: FW;
 
   /**
    * Maximum allowed pixel width of axis tick labels.
    *
    * __Default value:__ `160`.
    */
-  labelLimit?: number;
+  labelLimit?: N;
 
   /**
    * The offset of the legend label.
@@ -342,7 +367,7 @@ export interface BaseLegend {
    *
    * __Default value:__ `4`.
    */
-  labelOffset?: number;
+  labelOffset?: N;
 
   /**
    * The strategy to use for resolving overlap of labels in gradient legends. If `false`, no overlap reduction is attempted. If set to `true` (default) or `"parity"`, a strategy of removing every other label is used. If set to `"greedy"`, a linear scan of the labels is performed, removing any label that overlaps with the last visible label (this often works better for log-scaled axes).

--- a/types/spec/scale.d.ts
+++ b/types/spec/scale.d.ts
@@ -113,6 +113,7 @@ export interface SequentialScale extends BaseScale {
   type: 'sequential';
   range: RangeScheme;
   clamp?: boolean | SignalRef;
+  padding?: number | SignalRef;
   zero?: boolean | SignalRef;
   nice?: boolean | TimeInterval | SignalRef;
 }
@@ -129,6 +130,7 @@ export interface TimeScale extends BaseScale {
   type: 'time' | 'utc';
   range?: RangeScheme;
   clamp?: boolean | SignalRef;
+  padding?: number | SignalRef;
   nice?: boolean | TimeInterval | SignalRef;
 }
 export interface IdentityScale extends BaseScale {
@@ -146,6 +148,7 @@ export interface LinearScale extends BaseScale {
   range?: RangeScheme;
   interpolate?: ScaleInterpolate;
   clamp?: boolean | SignalRef;
+  padding?: number | SignalRef;
   nice?: boolean | number | SignalRef;
   zero?: boolean | SignalRef;
 }
@@ -155,6 +158,7 @@ export interface LogScale extends BaseScale {
   interpolate?: ScaleInterpolate;
   base?: number | SignalRef;
   clamp?: boolean | SignalRef;
+  padding?: number | SignalRef;
   nice?: boolean | number | SignalRef;
 }
 export interface PowScale extends BaseScale {
@@ -163,6 +167,7 @@ export interface PowScale extends BaseScale {
   range?: RangeScheme;
   interpolate?: ScaleInterpolate;
   clamp?: boolean | SignalRef;
+  padding?: number | SignalRef;
   nice?: boolean | number | SignalRef;
   zero?: boolean | SignalRef;
 }
@@ -171,6 +176,7 @@ export interface SqrtScale extends BaseScale {
   range?: RangeScheme;
   interpolate?: ScaleInterpolate;
   clamp?: boolean | SignalRef;
+  padding?: number | SignalRef;
   nice?: boolean | number | SignalRef;
   zero?: boolean | SignalRef;
 }
@@ -183,18 +189,21 @@ export interface QuantileScale extends BaseScale {
 export interface QuantizeScale extends BaseScale {
   type?: 'quantize';
   range?: RangeScheme;
+  padding?: number | SignalRef;
   nice?: boolean | number | SignalRef;
   zero?: boolean | SignalRef;
 }
 export interface ThresholdScale extends BaseScale {
   type?: 'threshold';
   range?: RangeScheme;
+  padding?: number | SignalRef;
   nice?: boolean | number | SignalRef;
   zero?: boolean | SignalRef;
 }
 export interface BinLinearScale extends BaseScale {
   type: 'bin-linear';
   range?: RangeScheme;
+  padding?: number | SignalRef;
   interpolate?: ScaleInterpolate;
 }
 export interface BinOrdinalScale extends BaseScale {

--- a/types/spec/signal.d.ts
+++ b/types/spec/signal.d.ts
@@ -18,5 +18,3 @@ export interface NewSignal extends BaseSignal {
   bind?: Binding;
 }
 export type Signal = NewSignal | PushSignal;
-
-export type WithSignal<T> = { [P in keyof T]: T[P] | SignalRef };

--- a/types/spec/util.d.ts
+++ b/types/spec/util.d.ts
@@ -4,5 +4,3 @@ export type Vector4<T> = [T, T, T, T];
 export type Vector5<T> = [T, T, T, T, T];
 export type Vector6<T> = [T, T, T, T, T, T];
 export type Vector7<T> = [T, T, T, T, T, T, T];
-
-export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

--- a/types/spec/values.d.ts
+++ b/types/spec/values.d.ts
@@ -1,0 +1,29 @@
+import {
+  NumericValueRef,
+  FontWeight,
+  StringValueRef,
+  FontWeightValueRef,
+  ColorValueRef,
+  AlignProperty,
+  Align,
+  AlignValueRef,
+  TextBaselineValueRef,
+  SymbolShape,
+  SymbolShapeValueRef,
+  Baseline,
+  TextBaseline,
+} from './encode';
+
+export type NumberValue = number | NumericValueRef;
+
+export type FontWeightValue = FontWeight | FontWeightValueRef;
+
+export type StringValue = string | StringValueRef;
+
+export type ColorValue = null | string | ColorValueRef;
+
+export type AlignValue = Align | AlignValueRef;
+
+export type TextBaselineValue = TextBaseline | TextBaselineValueRef;
+
+export type SymbolShapeValue = SymbolShape | SymbolShapeValueRef;

--- a/types/spec/values.d.ts
+++ b/types/spec/values.d.ts
@@ -12,6 +12,7 @@ import {
   SymbolShapeValueRef,
   Baseline,
   TextBaseline,
+  BooleanValueRef,
 } from './encode';
 
 export type NumberValue = number | NumericValueRef;
@@ -27,3 +28,5 @@ export type AlignValue = Align | AlignValueRef;
 export type TextBaselineValue = TextBaseline | TextBaselineValueRef;
 
 export type SymbolShapeValue = SymbolShape | SymbolShapeValueRef;
+
+export type BooleanValue = boolean | BooleanValueRef;

